{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "from feature_generation import seq2features, create_genomic_features\n",
    "import warnings\n",
    "from Bio import BiopythonDeprecationWarning\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=BiopythonDeprecationWarning)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_sequences(fasta_file):\n",
    "    # Count lines to estimate number of sequences (optional, for better progress bar)\n",
    "    with open(fasta_file, \"r\") as f:\n",
    "        n_seqs = sum(1 for line in f if line.startswith(\">\"))\n",
    "\n",
    "    with open(fasta_file, \"r\") as f:\n",
    "        entries = [\n",
    "            (name.split()[0], seq)\n",
    "            for name, seq in tqdm(\n",
    "                SimpleFastaParser(f), total=n_seqs, desc=\"Loading sequences\"\n",
    "            )\n",
    "        ]\n",
    "    return entries\n",
    "\n",
    "\n",
    "def generate_features(entries, is_dna=False, chunk_size=None):\n",
    "    if chunk_size is None:\n",
    "        # Original behavior\n",
    "        if is_dna:\n",
    "            features_df = create_genomic_features(\"your_dna.fasta\", file_format=\"fasta\")\n",
    "        else:\n",
    "            features_df = seq2features(entries, min_length=10, scaling=True)\n",
    "        return features_df\n",
    "    else:\n",
    "        features_list = []\n",
    "        for i in tqdm(range(0, len(entries), chunk_size), desc=\"Feature extraction\"):\n",
    "            chunk = entries[i : i + chunk_size]\n",
    "            features_chunk = seq2features(chunk, min_length=10, scaling=True)\n",
    "            features_list.append(features_chunk)\n",
    "        return pd.concat(features_list)\n",
    "\n",
    "\n",
    "def save_results(features_df, output_file):\n",
    "    features_df = features_df.reset_index().rename(columns={\"index\": \"id\"})\n",
    "    features_df = features_df.drop(columns=[\"md5\"])\n",
    "    features_df.to_parquet(output_file)\n",
    "    print(f\"Features saved to {output_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading sequences: 100%|██████████| 2621/2621 [00:00<00:00, 667350.86it/s]\n",
      "Feature extraction: 100%|██████████| 27/27 [00:00<00:00, 37.92it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Features saved to ../dataset/demonstration_samples/GCF_000009645.1/features_notebook.pa\n"
     ]
    }
   ],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    # Load sequences\n",
    "    entries = load_sequences(\n",
    "        \"../dataset/demonstration_samples/GCF_000009645.1/protein.faa\"\n",
    "    )\n",
    "\n",
    "    # Generate features\n",
    "    features_df = generate_features(entries, chunk_size=100)\n",
    "\n",
    "    # Save results\n",
    "    save_results(\n",
    "        features_df,\n",
    "        \"../dataset/demonstration_samples/GCF_000009645.1/features_notebook.pa\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "features_notebook = pd.read_parquet(\n",
    "    \"../dataset/demonstration_samples/GCF_000009645.1/features_notebook.pa\"\n",
    ")\n",
    "features_pipeline = pd.read_parquet(\n",
    "    \"../dataset/demonstration_samples/GCF_000009645.1/features.pa\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['id', 'PROSITE:ASX_HYDROXYL', 'PROSITE:PHOSPHOPANTETHEINE',\n",
      "       'PROSITE:EF_HAND_1', 'PROSITE:EGF_1', 'PROSITE:HOMEOBOX_1',\n",
      "       'PROSITE:ZINC_FINGER_C2H2_1', 'PROSITE:DEAD_ATP_HELICASE',\n",
      "       'PROSITE:RIBOSOMAL_S12', 'PROSITE:CYTOCHROME_P450',\n",
      "       ...\n",
      "       'RED_TRIPEP:SSA', 'RED_TRIPEP:SSC', 'RED_TRIPEP:SSE', 'RED_TRIPEP:SSF',\n",
      "       'RED_TRIPEP:SSG', 'RED_TRIPEP:SSH', 'RED_TRIPEP:SSK', 'RED_TRIPEP:SSL',\n",
      "       'RED_TRIPEP:SSP', 'RED_TRIPEP:SSS'],\n",
      "      dtype='object', length=1712)\n"
     ]
    }
   ],
   "source": [
    "print(features_notebook.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PROSITE:ASX_HYDROXYL</th>\n",
       "      <th>PROSITE:PHOSPHOPANTETHEINE</th>\n",
       "      <th>PROSITE:EF_HAND_1</th>\n",
       "      <th>PROSITE:EGF_1</th>\n",
       "      <th>PROSITE:HOMEOBOX_1</th>\n",
       "      <th>PROSITE:ZINC_FINGER_C2H2_1</th>\n",
       "      <th>PROSITE:DEAD_ATP_HELICASE</th>\n",
       "      <th>PROSITE:RIBOSOMAL_S12</th>\n",
       "      <th>PROSITE:CYTOCHROME_P450</th>\n",
       "      <th>PROSITE:CARBAMOYLTRANSFERASE</th>\n",
       "      <th>...</th>\n",
       "      <th>RED_TRIPEP:SSA</th>\n",
       "      <th>RED_TRIPEP:SSC</th>\n",
       "      <th>RED_TRIPEP:SSE</th>\n",
       "      <th>RED_TRIPEP:SSF</th>\n",
       "      <th>RED_TRIPEP:SSG</th>\n",
       "      <th>RED_TRIPEP:SSH</th>\n",
       "      <th>RED_TRIPEP:SSK</th>\n",
       "      <th>RED_TRIPEP:SSL</th>\n",
       "      <th>RED_TRIPEP:SSP</th>\n",
       "      <th>RED_TRIPEP:SSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 1711 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       PROSITE:ASX_HYDROXYL  PROSITE:PHOSPHOPANTETHEINE  PROSITE:EF_HAND_1  \\\n",
       "count                  True                        True               True   \n",
       "mean                   True                        True               True   \n",
       "std                    True                        True               True   \n",
       "min                    True                        True               True   \n",
       "25%                    True                        True               True   \n",
       "50%                    True                        True               True   \n",
       "75%                    True                        True               True   \n",
       "max                    True                        True               True   \n",
       "\n",
       "       PROSITE:EGF_1  PROSITE:HOMEOBOX_1  PROSITE:ZINC_FINGER_C2H2_1  \\\n",
       "count           True                True                        True   \n",
       "mean            True                True                        True   \n",
       "std             True                True                        True   \n",
       "min             True                True                        True   \n",
       "25%             True                True                        True   \n",
       "50%             True                True                        True   \n",
       "75%             True                True                        True   \n",
       "max             True                True                        True   \n",
       "\n",
       "       PROSITE:DEAD_ATP_HELICASE  PROSITE:RIBOSOMAL_S12  \\\n",
       "count                       True                   True   \n",
       "mean                        True                   True   \n",
       "std                         True                   True   \n",
       "min                         True                   True   \n",
       "25%                         True                   True   \n",
       "50%                         True                   True   \n",
       "75%                         True                   True   \n",
       "max                         True                   True   \n",
       "\n",
       "       PROSITE:CYTOCHROME_P450  PROSITE:CARBAMOYLTRANSFERASE  ...  \\\n",
       "count                     True                          True  ...   \n",
       "mean                      True                          True  ...   \n",
       "std                       True                          True  ...   \n",
       "min                       True                          True  ...   \n",
       "25%                       True                          True  ...   \n",
       "50%                       True                          True  ...   \n",
       "75%                       True                          True  ...   \n",
       "max                       True                          True  ...   \n",
       "\n",
       "       RED_TRIPEP:SSA  RED_TRIPEP:SSC  RED_TRIPEP:SSE  RED_TRIPEP:SSF  \\\n",
       "count            True            True            True            True   \n",
       "mean             True            True            True            True   \n",
       "std              True            True            True            True   \n",
       "min              True            True            True            True   \n",
       "25%              True            True            True            True   \n",
       "50%              True            True            True            True   \n",
       "75%              True            True            True            True   \n",
       "max              True            True            True            True   \n",
       "\n",
       "       RED_TRIPEP:SSG  RED_TRIPEP:SSH  RED_TRIPEP:SSK  RED_TRIPEP:SSL  \\\n",
       "count            True            True            True            True   \n",
       "mean             True            True            True            True   \n",
       "std              True            True            True            True   \n",
       "min              True            True            True            True   \n",
       "25%              True            True            True            True   \n",
       "50%              True            True            True            True   \n",
       "75%              True            True            True            True   \n",
       "max              True            True            True            True   \n",
       "\n",
       "       RED_TRIPEP:SSP  RED_TRIPEP:SSS  \n",
       "count            True            True  \n",
       "mean             True            True  \n",
       "std              True            True  \n",
       "min              True            True  \n",
       "25%              True            True  \n",
       "50%              True            True  \n",
       "75%              True            True  \n",
       "max              True            True  \n",
       "\n",
       "[8 rows x 1711 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_notebook.describe() == features_pipeline.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # write all column names to a file\n",
    "# with open(\"../document/feature_names.txt\", \"w\") as f:\n",
    "#     for col in features.columns:\n",
    "#         f.write(col + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lysis</th>\n",
       "      <th>tail</th>\n",
       "      <th>connector</th>\n",
       "      <th>dna_rna_and_nucleotide_metabolism</th>\n",
       "      <th>head_and_packaging</th>\n",
       "      <th>other</th>\n",
       "      <th>transcription_regulation</th>\n",
       "      <th>moron_auxiliary_metabolic_gene_and_host_takeover</th>\n",
       "      <th>unknown_function</th>\n",
       "      <th>integration_and_excision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1000.000000</td>\n",
       "      <td>1000.000000</td>\n",
       "      <td>1.000000e+03</td>\n",
       "      <td>1000.000000</td>\n",
       "      <td>1000.000000</td>\n",
       "      <td>1000.000000</td>\n",
       "      <td>1.000000e+03</td>\n",
       "      <td>1000.000000</td>\n",
       "      <td>1000.000000</td>\n",
       "      <td>1000.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.000639</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>5.119628e-07</td>\n",
       "      <td>0.858891</td>\n",
       "      <td>0.000345</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>4.713150e-10</td>\n",
       "      <td>0.040504</td>\n",
       "      <td>0.001123</td>\n",
       "      <td>0.000026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.000227</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>1.772102e-07</td>\n",
       "      <td>0.045859</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>2.363847e-10</td>\n",
       "      <td>0.010618</td>\n",
       "      <td>0.000625</td>\n",
       "      <td>0.000022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000110</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>5.832449e-08</td>\n",
       "      <td>0.667213</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>1.817702e-10</td>\n",
       "      <td>0.025386</td>\n",
       "      <td>0.000208</td>\n",
       "      <td>0.000007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.000482</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>4.039036e-07</td>\n",
       "      <td>0.831356</td>\n",
       "      <td>0.000234</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>2.818803e-10</td>\n",
       "      <td>0.034654</td>\n",
       "      <td>0.000714</td>\n",
       "      <td>0.000014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.000663</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>5.011168e-07</td>\n",
       "      <td>0.863870</td>\n",
       "      <td>0.000326</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>4.160402e-10</td>\n",
       "      <td>0.038846</td>\n",
       "      <td>0.001013</td>\n",
       "      <td>0.000020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.000802</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>6.282454e-07</td>\n",
       "      <td>0.890974</td>\n",
       "      <td>0.000425</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>6.113164e-10</td>\n",
       "      <td>0.044248</td>\n",
       "      <td>0.001379</td>\n",
       "      <td>0.000026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>0.001259</td>\n",
       "      <td>0.000256</td>\n",
       "      <td>1.206345e-06</td>\n",
       "      <td>0.961023</td>\n",
       "      <td>0.001633</td>\n",
       "      <td>0.000335</td>\n",
       "      <td>2.245561e-09</td>\n",
       "      <td>0.242920</td>\n",
       "      <td>0.007014</td>\n",
       "      <td>0.000239</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             lysis         tail     connector  \\\n",
       "count  1000.000000  1000.000000  1.000000e+03   \n",
       "mean      0.000639     0.000071  5.119628e-07   \n",
       "std       0.000227     0.000021  1.772102e-07   \n",
       "min       0.000110     0.000031  5.832449e-08   \n",
       "25%       0.000482     0.000058  4.039036e-07   \n",
       "50%       0.000663     0.000068  5.011168e-07   \n",
       "75%       0.000802     0.000080  6.282454e-07   \n",
       "max       0.001259     0.000256  1.206345e-06   \n",
       "\n",
       "       dna_rna_and_nucleotide_metabolism  head_and_packaging        other  \\\n",
       "count                        1000.000000         1000.000000  1000.000000   \n",
       "mean                            0.858891            0.000345     0.000060   \n",
       "std                             0.045859            0.000167     0.000031   \n",
       "min                             0.667213            0.000025     0.000012   \n",
       "25%                             0.831356            0.000234     0.000041   \n",
       "50%                             0.863870            0.000326     0.000055   \n",
       "75%                             0.890974            0.000425     0.000072   \n",
       "max                             0.961023            0.001633     0.000335   \n",
       "\n",
       "       transcription_regulation  \\\n",
       "count              1.000000e+03   \n",
       "mean               4.713150e-10   \n",
       "std                2.363847e-10   \n",
       "min                1.817702e-10   \n",
       "25%                2.818803e-10   \n",
       "50%                4.160402e-10   \n",
       "75%                6.113164e-10   \n",
       "max                2.245561e-09   \n",
       "\n",
       "       moron_auxiliary_metabolic_gene_and_host_takeover  unknown_function  \\\n",
       "count                                       1000.000000       1000.000000   \n",
       "mean                                           0.040504          0.001123   \n",
       "std                                            0.010618          0.000625   \n",
       "min                                            0.025386          0.000208   \n",
       "25%                                            0.034654          0.000714   \n",
       "50%                                            0.038846          0.001013   \n",
       "75%                                            0.044248          0.001379   \n",
       "max                                            0.242920          0.007014   \n",
       "\n",
       "       integration_and_excision  \n",
       "count               1000.000000  \n",
       "mean                   0.000026  \n",
       "std                    0.000022  \n",
       "min                    0.000007  \n",
       "25%                    0.000014  \n",
       "50%                    0.000020  \n",
       "75%                    0.000026  \n",
       "max                    0.000239  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pici_predictor.utilities import predict_function\n",
    "\n",
    "feature_df = pd.read_parquet(\"../dataset/protein_features_unique.pa\").sample(1000)\n",
    "out_dir = \"../results/demonstration/GCF_000009645.1/predicted_prob.csv\"\n",
    "model_path = \"../models/best_configs\"\n",
    "predicted_prob = predict_function(feature_df, out_dir, model_path)\n",
    "predicted_prob.describe()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
