{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pharokka_cds_functions = pd.read_csv('../results/pharokka_output/pharokka_cds_functions.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>Count</th>\n",
       "      <th>contig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDS</td>\n",
       "      <td>26</td>\n",
       "      <td>STAU002.0321.00050.C001.PICI.TypeB.variant0002...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>connector</td>\n",
       "      <td>0</td>\n",
       "      <td>STAU002.0321.00050.C001.PICI.TypeB.variant0002...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DNA, RNA and nucleotide metabolism</td>\n",
       "      <td>3</td>\n",
       "      <td>STAU002.0321.00050.C001.PICI.TypeB.variant0002...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>head and packaging</td>\n",
       "      <td>2</td>\n",
       "      <td>STAU002.0321.00050.C001.PICI.TypeB.variant0002...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>integration and excision</td>\n",
       "      <td>1</td>\n",
       "      <td>STAU002.0321.00050.C001.PICI.TypeB.variant0002...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Description  Count  \\\n",
       "0                                 CDS     26   \n",
       "1                           connector      0   \n",
       "2  DNA, RNA and nucleotide metabolism      3   \n",
       "3                  head and packaging      2   \n",
       "4            integration and excision      1   \n",
       "\n",
       "                                              contig  \n",
       "0  STAU002.0321.00050.C001.PICI.TypeB.variant0002...  \n",
       "1  STAU002.0321.00050.C001.PICI.TypeB.variant0002...  \n",
       "2  STAU002.0321.00050.C001.PICI.TypeB.variant0002...  \n",
       "3  STAU002.0321.00050.C001.PICI.TypeB.variant0002...  \n",
       "4  STAU002.0321.00050.C001.PICI.TypeB.variant0002...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pharokka_cds_functions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26843\n",
      "ID: WLQYMNQG_CDS_0001\n",
      "Name: WLQYMNQG_CDS_0001\n",
      "Description: WLQYMNQG_CDS_0001 integrase\n",
      "Number of features: 0\n",
      "Seq('MWVREITKNNSTAYRYLERYTDPLTGKYKTVSVTRNKNNVRSQKDAQLELNKII...IGS')\n"
     ]
    }
   ],
   "source": [
    "phanotate_faa = list(SeqIO.parse(\"../results/pharokka_output/phanotate.faa\", \"fasta\"))\n",
    "print(len(phanotate_faa))\n",
    "print(phanotate_faa[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(26843, 2)\n",
      "26843\n",
      "104\n"
     ]
    }
   ],
   "source": [
    "gene_ids = []\n",
    "functions = []\n",
    "\n",
    "for seq in phanotate_faa:\n",
    "    seq_id = seq.id\n",
    "    function = seq.description.replace(seq_id, '').strip()\n",
    "    \n",
    "    gene_ids.append(seq_id)\n",
    "    functions.append(function)\n",
    "\n",
    "phanotate_functions = pd.DataFrame({\n",
    "    'gene_id': gene_ids,\n",
    "    'function': functions\n",
    "})\n",
    "\n",
    "print(phanotate_functions.shape)\n",
    "print(len(phanotate_functions[\"gene_id\"].unique()))\n",
    "print(len(phanotate_functions[\"function\"].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WLQYMNQG_CDS_0001</td>\n",
       "      <td>integrase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WLQYMNQG_CDS_0002</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WLQYMNQG_CDS_0003</td>\n",
       "      <td>transcriptional repressor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WLQYMNQG_CDS_0004</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WLQYMNQG_CDS_0005</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_id                   function\n",
       "0  WLQYMNQG_CDS_0001                  integrase\n",
       "1  WLQYMNQG_CDS_0002       hypothetical protein\n",
       "2  WLQYMNQG_CDS_0003  transcriptional repressor\n",
       "3  WLQYMNQG_CDS_0004       hypothetical protein\n",
       "4  WLQYMNQG_CDS_0005       hypothetical protein"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phanotate_functions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "phanotate_functions[\"function\"].value_counts().to_csv(\"../results/pharokka_output/phanotate_functions_counts.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(26843, 43)\n",
      "26843\n",
      "1435\n"
     ]
    }
   ],
   "source": [
    "pharokka_cds_final_merged_output = pd.read_csv(\"../results/pharokka_output/pharokka_cds_final_merged_output.tsv\", sep='\\t')\n",
    "print(pharokka_cds_final_merged_output.shape)\n",
    "print(len(pharokka_cds_final_merged_output[\"gene\"].unique()))\n",
    "print(len(pharokka_cds_final_merged_output[\"contig\"].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>frame</th>\n",
       "      <th>contig</th>\n",
       "      <th>score</th>\n",
       "      <th>mmseqs_phrog</th>\n",
       "      <th>mmseqs_alnScore</th>\n",
       "      <th>mmseqs_seqIdentity</th>\n",
       "      <th>mmseqs_eVal</th>\n",
       "      <th>...</th>\n",
       "      <th>CARD_eVal</th>\n",
       "      <th>CARD_species</th>\n",
       "      <th>ARO_Accession</th>\n",
       "      <th>CARD_short_name</th>\n",
       "      <th>Protein_Accession</th>\n",
       "      <th>DNA_Accession</th>\n",
       "      <th>AMR_Gene_Family</th>\n",
       "      <th>Drug_Class</th>\n",
       "      <th>Resistance_Mechanism</th>\n",
       "      <th>transl_table</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WLQYMNQG_CDS_0001</td>\n",
       "      <td>1214</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>STAU002.0321.00050.C001.PICI.TypeB.variant0002...</td>\n",
       "      <td>-3.319236e+11</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WLQYMNQG_CDS_0002</td>\n",
       "      <td>1386</td>\n",
       "      <td>1231</td>\n",
       "      <td>-</td>\n",
       "      <td>STAU002.0321.00050.C001.PICI.TypeB.variant0002...</td>\n",
       "      <td>-1.120867e-01</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WLQYMNQG_CDS_0003</td>\n",
       "      <td>2203</td>\n",
       "      <td>1478</td>\n",
       "      <td>-</td>\n",
       "      <td>STAU002.0321.00050.C001.PICI.TypeB.variant0002...</td>\n",
       "      <td>-1.596539e+09</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WLQYMNQG_CDS_0004</td>\n",
       "      <td>2321</td>\n",
       "      <td>2223</td>\n",
       "      <td>-</td>\n",
       "      <td>STAU002.0321.00050.C001.PICI.TypeB.variant0002...</td>\n",
       "      <td>-1.078696e+00</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WLQYMNQG_CDS_0005</td>\n",
       "      <td>2375</td>\n",
       "      <td>2587</td>\n",
       "      <td>+</td>\n",
       "      <td>STAU002.0321.00050.C001.PICI.TypeB.variant0002...</td>\n",
       "      <td>-3.138978e+02</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26838</th>\n",
       "      <td>WLQYMNQG_CDS_26839</td>\n",
       "      <td>6506</td>\n",
       "      <td>6393</td>\n",
       "      <td>-</td>\n",
       "      <td>STSI001.0321.00006.C001.PICI.TypeB.variant0001...</td>\n",
       "      <td>-9.468694e-02</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26839</th>\n",
       "      <td>WLQYMNQG_CDS_26840</td>\n",
       "      <td>6608</td>\n",
       "      <td>7588</td>\n",
       "      <td>+</td>\n",
       "      <td>STSI001.0321.00006.C001.PICI.TypeB.variant0001...</td>\n",
       "      <td>-5.211945e+06</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26840</th>\n",
       "      <td>WLQYMNQG_CDS_26841</td>\n",
       "      <td>7632</td>\n",
       "      <td>7817</td>\n",
       "      <td>+</td>\n",
       "      <td>STSI001.0321.00006.C001.PICI.TypeB.variant0001...</td>\n",
       "      <td>-4.271133e+00</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26841</th>\n",
       "      <td>WLQYMNQG_CDS_26842</td>\n",
       "      <td>7804</td>\n",
       "      <td>8370</td>\n",
       "      <td>+</td>\n",
       "      <td>STSI001.0321.00006.C001.PICI.TypeB.variant0001...</td>\n",
       "      <td>-6.485262e+06</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26842</th>\n",
       "      <td>WLQYMNQG_CDS_26843</td>\n",
       "      <td>8372</td>\n",
       "      <td>9604</td>\n",
       "      <td>+</td>\n",
       "      <td>STSI001.0321.00006.C001.PICI.TypeB.variant0001...</td>\n",
       "      <td>-1.080020e+11</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>No_MMseqs</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26843 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene  start  stop frame  \\\n",
       "0       WLQYMNQG_CDS_0001   1214     3     -   \n",
       "1       WLQYMNQG_CDS_0002   1386  1231     -   \n",
       "2       WLQYMNQG_CDS_0003   2203  1478     -   \n",
       "3       WLQYMNQG_CDS_0004   2321  2223     -   \n",
       "4       WLQYMNQG_CDS_0005   2375  2587     +   \n",
       "...                   ...    ...   ...   ...   \n",
       "26838  WLQYMNQG_CDS_26839   6506  6393     -   \n",
       "26839  WLQYMNQG_CDS_26840   6608  7588     +   \n",
       "26840  WLQYMNQG_CDS_26841   7632  7817     +   \n",
       "26841  WLQYMNQG_CDS_26842   7804  8370     +   \n",
       "26842  WLQYMNQG_CDS_26843   8372  9604     +   \n",
       "\n",
       "                                                  contig         score  \\\n",
       "0      STAU002.0321.00050.C001.PICI.TypeB.variant0002... -3.319236e+11   \n",
       "1      STAU002.0321.00050.C001.PICI.TypeB.variant0002... -1.120867e-01   \n",
       "2      STAU002.0321.00050.C001.PICI.TypeB.variant0002... -1.596539e+09   \n",
       "3      STAU002.0321.00050.C001.PICI.TypeB.variant0002... -1.078696e+00   \n",
       "4      STAU002.0321.00050.C001.PICI.TypeB.variant0002... -3.138978e+02   \n",
       "...                                                  ...           ...   \n",
       "26838  STSI001.0321.00006.C001.PICI.TypeB.variant0001... -9.468694e-02   \n",
       "26839  STSI001.0321.00006.C001.PICI.TypeB.variant0001... -5.211945e+06   \n",
       "26840  STSI001.0321.00006.C001.PICI.TypeB.variant0001... -4.271133e+00   \n",
       "26841  STSI001.0321.00006.C001.PICI.TypeB.variant0001... -6.485262e+06   \n",
       "26842  STSI001.0321.00006.C001.PICI.TypeB.variant0001... -1.080020e+11   \n",
       "\n",
       "      mmseqs_phrog mmseqs_alnScore mmseqs_seqIdentity mmseqs_eVal  ...  \\\n",
       "0        No_MMseqs       No_MMseqs          No_MMseqs   No_MMseqs  ...   \n",
       "1        No_MMseqs       No_MMseqs          No_MMseqs   No_MMseqs  ...   \n",
       "2        No_MMseqs       No_MMseqs          No_MMseqs   No_MMseqs  ...   \n",
       "3        No_MMseqs       No_MMseqs          No_MMseqs   No_MMseqs  ...   \n",
       "4        No_MMseqs       No_MMseqs          No_MMseqs   No_MMseqs  ...   \n",
       "...            ...             ...                ...         ...  ...   \n",
       "26838    No_MMseqs       No_MMseqs          No_MMseqs   No_MMseqs  ...   \n",
       "26839    No_MMseqs       No_MMseqs          No_MMseqs   No_MMseqs  ...   \n",
       "26840    No_MMseqs       No_MMseqs          No_MMseqs   No_MMseqs  ...   \n",
       "26841    No_MMseqs       No_MMseqs          No_MMseqs   No_MMseqs  ...   \n",
       "26842    No_MMseqs       No_MMseqs          No_MMseqs   No_MMseqs  ...   \n",
       "\n",
       "      CARD_eVal CARD_species ARO_Accession CARD_short_name Protein_Accession  \\\n",
       "0           NaN          NaN           NaN             NaN               NaN   \n",
       "1           NaN          NaN           NaN             NaN               NaN   \n",
       "2           NaN          NaN           NaN             NaN               NaN   \n",
       "3           NaN          NaN           NaN             NaN               NaN   \n",
       "4           NaN          NaN           NaN             NaN               NaN   \n",
       "...         ...          ...           ...             ...               ...   \n",
       "26838       NaN          NaN           NaN             NaN               NaN   \n",
       "26839       NaN          NaN           NaN             NaN               NaN   \n",
       "26840       NaN          NaN           NaN             NaN               NaN   \n",
       "26841       NaN          NaN           NaN             NaN               NaN   \n",
       "26842       NaN          NaN           NaN             NaN               NaN   \n",
       "\n",
       "      DNA_Accession AMR_Gene_Family Drug_Class Resistance_Mechanism  \\\n",
       "0               NaN             NaN        NaN                  NaN   \n",
       "1               NaN             NaN        NaN                  NaN   \n",
       "2               NaN             NaN        NaN                  NaN   \n",
       "3               NaN             NaN        NaN                  NaN   \n",
       "4               NaN             NaN        NaN                  NaN   \n",
       "...             ...             ...        ...                  ...   \n",
       "26838           NaN             NaN        NaN                  NaN   \n",
       "26839           NaN             NaN        NaN                  NaN   \n",
       "26840           NaN             NaN        NaN                  NaN   \n",
       "26841           NaN             NaN        NaN                  NaN   \n",
       "26842           NaN             NaN        NaN                  NaN   \n",
       "\n",
       "      transl_table  \n",
       "0               11  \n",
       "1               11  \n",
       "2               11  \n",
       "3               11  \n",
       "4               11  \n",
       "...            ...  \n",
       "26838           11  \n",
       "26839           11  \n",
       "26840           11  \n",
       "26841           11  \n",
       "26842           11  \n",
       "\n",
       "[26843 rows x 43 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pharokka_cds_final_merged_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for gene_id in phanotate_functions, find the pici_id in pharokka_cds_final_merged_output, add pici_id to phanotate_functions; pici_id is in the contig column of pharokka_cds_final_merged_output\n",
    "phanotate_functions[\"pici_id\"] = phanotate_functions[\"gene_id\"].map(pharokka_cds_final_merged_output.set_index(\"gene\")[\"contig\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(26843, 3)\n",
      "26843\n",
      "1435\n"
     ]
    }
   ],
   "source": [
    "print(phanotate_functions.shape)\n",
    "print(len(phanotate_functions[\"gene_id\"].unique()))\n",
    "print(len(phanotate_functions[\"pici_id\"].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "phanotate_functions.to_csv(\"../results/pharokka_output/phanotate_functions.tsv\", sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
